### PR TITLE
ARROW-13110: [C++] Deadlock can happen when using BackgroundGenerator without transferring callbacks

### PR DESCRIPTION
**Pull request by westonpace**

### CODE REVIEW

File: cpp/src/arrow/util/async_generator.h [hunk 1]
```diff
@@ -20,6 +20,7 @@
 #include <cassert>
 #include <deque>
 #include <queue>
+#include <thread>
```
Comment (westonpace) [inline cpp/src/arrow/util/async_generator.h:23]: This is unfortunate but we could easily add a method for getting the current thread id into thread_pool.h once ARROW-12878 is merged.

**Comment by westonpace** [comment]
@lidavidm Do you see any harm that could be caused by this transfer?

**Comment by westonpace** [comment]
Related: ARROW-13109 should allow the producer to place restrictions on callbacks.

**Comment by lidavidm** [comment]
You mean the one in OneShotFragment? I think that's fine.

Should we add the thread pool as a parameter to MakeBackgroundGenerator? So that you can't forget. (I guess that'd create some churn once we have ARROW-13109 since it would become redundant again.)

This looks like it was a fun issue to debug…

**Review by bkietz — APPROVED** [review]
LGTM

**Pull request closed without merging**